src/track/beats.cpp

```diff
@@ -53,3 +190,676 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+
+    track::io::Beat beat;
+    beat.set_frame_position(frame);
+
+    // it points at the first occurrence of beat or the next largest beat
+    BeatList::const_iterator it =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), beat, BeatLessThan);
+
+    // If the position is within 1/10th of a second of the next or previous
+    // beat, pretend we are on that beat.
+    const double kFrameEpsilon = 0.1 * m_iSampleRate;
+
+    // Back-up by one.
+    if (it != m_beats.begin()) {
+        --it;
+    }
+
+    // Scan forward to find whether we are on a beat.
+    BeatList::const_iterator on_beat = m_beats.cend();
+    BeatList::const_iterator previous_beat = m_beats.cend();
+    BeatList::const_iterator next_beat = m_beats.cend();
+    for (; it != m_beats.end(); ++it) {
+        qint32 delta = it->frame_position() - beat.frame_position();
+
+        // We are "on" this beat.
+        if (abs(delta) < kFrameEpsilon) {
+            on_beat = it;
+            break;
+        }
+
+        if (delta < 0) {
+            // If we are not on the beat and delta < 0 then this beat comes
+            // before our current position.
+            previous_beat = it;
+        } else {
+            // If we are past the beat and we aren't on it then this beat comes
+            // after our current position.
+            next_beat = it;
+            // Stop because we have everything we need now.
+            break;
+        }
+    }
+
+    // If we are within epsilon samples of a beat then the immediately next and
+    // previous beats are the beat we are on.
+    if (on_beat != m_beats.end()) {
+        previous_beat = on_beat;
+        next_beat = on_beat + 1;
+    }
+
+    *pPrevBeatFrame = -1;
+    *pNextBeatFrame = -1;
+
+    for (; next_beat != m_beats.end(); ++next_beat) {
+        if (!next_beat->enabled()) {
+            continue;
+        }
+        *pNextBeatFrame = next_beat->frame_position();
+        break;
+    }
+    if (previous_beat != m_beats.end()) {
+        for (; true; --previous_beat) {
+            if (previous_beat->enabled()) {
+                *pPrevBeatFrame = previous_beat->frame_position();
+                break;
+            }
+
+            // Don't step before the start of the list.
+            if (previous_beat == m_beats.begin()) {
+                break;
+            }
+        }
+    }
+    return *pPrevBeatFrame != -1 && *pNextBeatFrame != -1;
+}
+
+double Beats::findClosestBeatNew(FrameNum frames) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid()) {
+        return -1;
+    }
+    double prevBeat;
+    double nextBeat;
+    findPrevNextBeatsNew(frames, &prevBeat, &nextBeat);
+    if (prevBeat == -1) {
+        // If both values are -1, we correctly return -1.
+        return nextBeat;
+    } else if (nextBeat == -1) {
+        return prevBeat;
+    }
+    return (nextBeat - frames > frames - prevBeat) ? prevBeat : nextBeat;
+}
+
+double Beats::findNthBeatNew(FrameNum frame, int n) const {
```

Comment:
Are all the possible paths in this function covered by tests? I am unable to decide if this complex code is correct or not.